src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs

```diff
@@ -0,0 +1,299 @@
+                    // based on most secure scheme to least secure scheme ordering.
+                    if (!Interop.WinHttp.WinHttpQueryAuthSchemes(
+                        state.RequestHandle,
+                        out supportedSchemes,
+                        out firstSchemeIgnored,
+                        out authTarget))
+                    {
+                        WinHttpException.ThrowExceptionUsingLastError();
+                    }
+
+                    // WinHTTP returns the proper authTarget based on the status code (401, 407).
+                    // But we can validate with assert.
+                    Debug.Assert(authTarget == Interop.WinHttp.WINHTTP_AUTH_TARGET_PROXY);
+
+                    proxyAuthScheme = ChooseAuthScheme(supportedSchemes);
+                    state.RetryRequest = true;
+                    break;
+
+                default:
+                    if (state.PreAuthenticate && serverAuthScheme != 0)
+                    {
+                        SaveServerCredentialsToCache(uri, serverAuthScheme, state.ServerCredentials);
+                    }
+                    break;
+            }
+        }
+
+        public void PreAuthenticateRequest(WinHttpRequestState state, uint proxyAuthScheme)
+        {
+            // Set proxy credentials if we have them.
+            // If a proxy authentication challenge was responded to, reset
+            // those credentials before each SendRequest, because the proxy  
+            // may require re-authentication after responding to a 401 or  
+            // to a redirect. If you don't, you can get into a 
+            // 407-401-407-401- loop.
+            if (proxyAuthScheme != 0)
+            {
+                SetWinHttpCredential(
+                    state.RequestHandle,
+                    state.Proxy == null ? state.DefaultProxyCredentials : state.Proxy.Credentials,
+                    state.RequestMessage.RequestUri,
+                    proxyAuthScheme,
+                    Interop.WinHttp.WINHTTP_AUTH_TARGET_PROXY);
+            }
+
+            // Apply pre-authentication headers for server authentication?
+            if (state.PreAuthenticate)
+            {
+                uint authScheme;
+                NetworkCredential serverCredentials;
+                if (GetServerCredentialsFromCache(
+                    state.RequestMessage.RequestUri,
+                    out authScheme,
+                    out serverCredentials))
+                {
+                    SetWinHttpCredential(
+                        state.RequestHandle,
+                        serverCredentials,
+                        state.RequestMessage.RequestUri,
+                        authScheme,
+                        Interop.WinHttp.WINHTTP_AUTH_TARGET_SERVER);
+                    state.LastStatusCode = HttpStatusCode.Unauthorized; // Remember we already set the creds.
+                }
```

Comment:
There is no missing code. There is nothing to log since there is not an "error" condition.

 If there is no cached cred (including the auth scheme to use), then the request goes out on the wire with no "Authorization" request header on it. Then a 401 will probably happen. Then the HTTP stack will be able to determine the auth scheme to use (based on the "WWW-Authenticate" response header. Then, that would be cached for later.